`invalid memory address or nil pointer dereference` when trying to create `RedisCluster`

Hello!

This `RedisCluster` manifest:

```
---

apiVersion: redis.redis.opstreelabs.in/v1beta1
kind: RedisCluster
metadata:
  name: redis-crash
  namespace: redis
spec:
  clusterSize: 3
  kubernetesConfig:
    image: quay.io/opstree/redis:v6.2.5
    imagePullPolicy: IfNotPresent
    resources:
      requests:
        cpu: 1000m
        memory: 1Gi
      limits:
        cpu: 1000m
        memory: 1Gi
    redisSecret:
      name: redis-crash
      key: password
  redisExporter:
    enabled: true
    image: quay.io/opstree/redis-exporter:1.0
  storage:
    volumeClaimTemplate:
      spec:
        storageClassName: foobar-csi
        accessModes: ["ReadWriteOnce"]
        resources:
          requests:
            storage: 5Gi
  redisLeader:
    affinity:
      podAntiAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
                - key: app
                  operator: In
                  values:
                    - redis-cluster-leader
            topologyKey: "kubernetes.io/hostname"
  redisFollower:
    affinity:
      podAntiAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
                - key: app
                  operator: In
                  values:
                    - redis-cluster-follower
            topologyKey: "kubernetes.io/hostname"
```

crashes `redis-operator` (`0.8.0`, installed via the Helm chart):

```
E0921 16:27:32.516068       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 305 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1672800, 0x23e20c0)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/runtime/runtime.go:74 +0xa6
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/runtime/runtime.go:48 +0x89
panic(0x1672800, 0x23e20c0)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9
redis-operator/k8sutils.enableRedisMonitoring(0xc00023cce0, 0x1c, 0xc000313440, 0xc, 0xc00062b4b0, 0xc0005e53e0, 0x22, 0x0, 0x0, 0x0, ...)
	/workspace/k8sutils/statefulset.go:192 +0x14f
redis-operator/k8sutils.generateContainerDef(0xc000652a80, 0x13, 0xc00023cce0, 0x1c, 0xc000313440, 0xc, 0xc00062b4b0, 0xc0005e53e0, 0x22, 0x0, ...)
	/workspace/k8sutils/statefulset.go:180 +0x498
redis-operator/k8sutils.generateStateFulSetsDef(0xc000652a80, 0x13, 0x0, 0x0, 0xc0003133e9, 0x5, 0x0, 0x0, 0x0, 0x0, ...)
	/workspace/k8sutils/statefulset.go:108 +0xd7
redis-operator/k8sutils.CreateOrUpdateStateFul(0xc0003133e9, 0x5, 0xc000652a80, 0x13, 0x0, 0x0, 0xc0003133e9, 0x5, 0x0, 0x0, ...)
	/workspace/k8sutils/statefulset.go:54 +0x2b8
redis-operator/k8sutils.RedisClusterSTS.CreateRedisClusterSetup(0x1836536, 0x6, 0x0, 0xc00071ea00, 0xc000364900, 0x0, 0x0)
	/workspace/k8sutils/redis-cluster.go:136 +0x8b8
redis-operator/k8sutils.CreateRedisLeader(0xc000364900, 0xc000364900, 0x1a52700)
	/workspace/k8sutils/redis-cluster.go:82 +0x65
redis-operator/controllers.(*RedisClusterReconciler).Reconcile(0xc0004dfb60, 0x1a2dec0, 0xc000655590, 0xc0003133e9, 0x5, 0xc0003133d4, 0xc, 0xc000655590, 0x40a27f, 0xc000030000, ...)
	/workspace/controllers/rediscluster_controller.go:60 +0x33b
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00034d400, 0x1a2de00, 0xc0000c4280, 0x16c9e20, 0xc000238640)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:263 +0x317
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00034d400, 0x1a2de00, 0xc0000c4280, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:235 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1(0x1a2de00, 0xc0000c4280)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:198 +0x4a
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185 +0x37
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000107750)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000733f50, 0x19f76a0, 0xc0006554a0, 0xc0000c4201, 0xc0004bd320)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:156 +0xad
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000107750, 0x3b9aca00, 0x0, 0x1, 0xc0004bd320)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext(0x1a2de00, 0xc0000c4280, 0xc00062b3c0, 0x3b9aca00, 0x0, 0xc000074501)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185 +0xa6
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(0x1a2de00, 0xc0000c4280, 0xc00062b3c0, 0x3b9aca00)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:99 +0x57
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:195 +0x4e7
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x15371ef]

goroutine 305 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/runtime/runtime.go:55 +0x10c
panic(0x1672800, 0x23e20c0)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9
redis-operator/k8sutils.enableRedisMonitoring(0xc00023cce0, 0x1c, 0xc000313440, 0xc, 0xc00062b4b0, 0xc0005e53e0, 0x22, 0x0, 0x0, 0x0, ...)
	/workspace/k8sutils/statefulset.go:192 +0x14f
redis-operator/k8sutils.generateContainerDef(0xc000652a80, 0x13, 0xc00023cce0, 0x1c, 0xc000313440, 0xc, 0xc00062b4b0, 0xc0005e53e0, 0x22, 0x0, ...)
	/workspace/k8sutils/statefulset.go:180 +0x498
redis-operator/k8sutils.generateStateFulSetsDef(0xc000652a80, 0x13, 0x0, 0x0, 0xc0003133e9, 0x5, 0x0, 0x0, 0x0, 0x0, ...)
	/workspace/k8sutils/statefulset.go:108 +0xd7
redis-operator/k8sutils.CreateOrUpdateStateFul(0xc0003133e9, 0x5, 0xc000652a80, 0x13, 0x0, 0x0, 0xc0003133e9, 0x5, 0x0, 0x0, ...)
	/workspace/k8sutils/statefulset.go:54 +0x2b8
redis-operator/k8sutils.RedisClusterSTS.CreateRedisClusterSetup(0x1836536, 0x6, 0x0, 0xc00071ea00, 0xc000364900, 0x0, 0x0)
	/workspace/k8sutils/redis-cluster.go:136 +0x8b8
redis-operator/k8sutils.CreateRedisLeader(0xc000364900, 0xc000364900, 0x1a52700)
	/workspace/k8sutils/redis-cluster.go:82 +0x65
redis-operator/controllers.(*RedisClusterReconciler).Reconcile(0xc0004dfb60, 0x1a2dec0, 0xc000655590, 0xc0003133e9, 0x5, 0xc0003133d4, 0xc, 0xc000655590, 0x40a27f, 0xc000030000, ...)
	/workspace/controllers/rediscluster_controller.go:60 +0x33b
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00034d400, 0x1a2de00, 0xc0000c4280, 0x16c9e20, 0xc000238640)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:263 +0x317
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00034d400, 0x1a2de00, 0xc0000c4280, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:235 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1(0x1a2de00, 0xc0000c4280)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:198 +0x4a
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185 +0x37
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000107750)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc000733f50, 0x19f76a0, 0xc0006554a0, 0xc0000c4201, 0xc0004bd320)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:156 +0xad
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000107750, 0x3b9aca00, 0x0, 0x1, 0xc0004bd320)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext(0x1a2de00, 0xc0000c4280, 0xc00062b3c0, 0x3b9aca00, 0x0, 0xc000074501)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185 +0xa6
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(0x1a2de00, 0xc0000c4280, 0xc00062b3c0, 0x3b9aca00)
	/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:99 +0x57
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:195 +0x4e7
```

`redis-operator` is stuck in `CrashLoopBackOff` afterwards until the `RedisCluster` resource is removed.

What could went wrong here?

--- a/k8sutils/redis-standalone.go
+++ b/k8sutils/redis-standalone.go
@@ -59,7 +59,6 @@ func generateRedisStandaloneParams(cr *redisv1beta1.Redis) statefulSetParameters
 		PriorityClassName: cr.Spec.PriorityClassName,
 		Affinity:          cr.Spec.Affinity,
 		Tolerations:       cr.Spec.Tolerations,
-		EnableMetrics:     cr.Spec.RedisExporter.Enabled,
 	}
 	if cr.Spec.KubernetesConfig.ImagePullSecrets != nil {
 		res.ImagePullSecrets = cr.Spec.KubernetesConfig.ImagePullSecrets
@@ -70,6 +69,10 @@ func generateRedisStandaloneParams(cr *redisv1beta1.Redis) statefulSetParameters
 	if cr.Spec.RedisConfig != nil {
 		res.ExternalConfig = cr.Spec.RedisConfig.AdditionalRedisConfig
 	}
+	if cr.Spec.RedisExporter != nil {
+		res.EnableMetrics = cr.Spec.RedisExporter.Enabled
+
+	}
 	return res
 }
 
@@ -78,13 +81,10 @@ func generateRedisStandaloneContainerParams(cr *redisv1beta1.Redis) containerPar
 	trueProperty := true
 	falseProperty := false
 	containerProp := containerParameters{
-		Role:                         "standalone",
-		Image:                        cr.Spec.KubernetesConfig.Image,
-		ImagePullPolicy:              cr.Spec.KubernetesConfig.ImagePullPolicy,
-		Resources:                    cr.Spec.KubernetesConfig.Resources,
-		RedisExporterImage:           cr.Spec.RedisExporter.Image,
-		RedisExporterImagePullPolicy: cr.Spec.RedisExporter.ImagePullPolicy,
-		RedisExporterResources:       cr.Spec.RedisExporter.Resources,
+		Role:            "standalone",
+		Image:           cr.Spec.KubernetesConfig.Image,
+		ImagePullPolicy: cr.Spec.KubernetesConfig.ImagePullPolicy,
+		Resources:       cr.Spec.KubernetesConfig.Resources,
 	}
 	if cr.Spec.KubernetesConfig.ExistingPasswordSecret != nil {
 		containerProp.EnabledPassword = &trueProperty
@@ -93,8 +93,15 @@ func generateRedisStandaloneContainerParams(cr *redisv1beta1.Redis) containerPar
 	} else {
 		containerProp.EnabledPassword = &falseProperty
 	}
-	if cr.Spec.RedisExporter.EnvVars != nil {
-		containerProp.RedisExporterEnv = cr.Spec.RedisExporter.EnvVars
+	if cr.Spec.RedisExporter != nil {
+		containerProp.RedisExporterImage = cr.Spec.RedisExporter.Image
+		containerProp.RedisExporterImagePullPolicy = cr.Spec.RedisExporter.ImagePullPolicy
+		containerProp.RedisExporterResources = cr.Spec.RedisExporter.Resources
+
+		if cr.Spec.RedisExporter.EnvVars != nil {
+			containerProp.RedisExporterEnv = cr.Spec.RedisExporter.EnvVars
+		}
+
 	}
 	if cr.Spec.Storage != nil {
 		containerProp.PersistenceEnabled = &trueProperty

--- a/k8sutils/statefulset.go
+++ b/k8sutils/statefulset.go
@@ -170,11 +170,13 @@ func generateContainerDef(name string, containerParams containerParameters, enab
 			Image:           containerParams.Image,
 			ImagePullPolicy: containerParams.ImagePullPolicy,
 			Env:             getEnvironmentVariables(containerParams.Role, containerParams.EnabledPassword, containerParams.SecretName, containerParams.SecretKey, containerParams.PersistenceEnabled, containerParams.RedisExporterEnv),
-			Resources:       *containerParams.Resources,
 			ReadinessProbe:  getProbeInfo(),
 			LivenessProbe:   getProbeInfo(),
 			VolumeMounts:    getVolumeMount(name, containerParams.PersistenceEnabled, externalConfig),
 		},
+	}
+	if containerParams.Resources != nil {
+		containerDefinition[0].Resources = *containerParams.Resources
 	}
 	if enableMetrics {
 		containerDefinition = append(containerDefinition, enableRedisMonitoring(containerParams))
@@ -189,7 +191,9 @@ func enableRedisMonitoring(params containerParameters) corev1.Container {
 		Image:           params.RedisExporterImage,
 		ImagePullPolicy: params.RedisExporterImagePullPolicy,
 		Env:             getEnvironmentVariables(params.Role, params.EnabledPassword, params.SecretName, params.SecretKey, params.PersistenceEnabled, params.RedisExporterEnv),
-		Resources:       *params.RedisExporterResources,
+	}
+	if params.RedisExporterResources != nil {
+		exporterDefinition.Resources = *params.RedisExporterResources
 	}
 	return exporterDefinition
 }

